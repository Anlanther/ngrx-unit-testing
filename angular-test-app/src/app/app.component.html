<div class="container">
  <div class="row header">
    @if (userName !== '') {
    <h1>Hello, {{ userName }}!</h1>
    } @else {
    <h1>Hello!</h1>
    }
    <select
      class="form-select"
      (change)="userSelected()"
      [formControl]="userForm"
    >
      <option selected>Select User:</option>
      @for (user of allUsers$ | async; track user.name) {
      <option value="{{ user.name }}">{{ user.name }}</option>
      }
    </select>
  </div>
  <div class="row content">
    <div class="col">
      <h2>Favourite Stories</h2>
      <ul class="list-group">
        @for (story of favouritedStories$ | async; track story.title) {
        <li class="list-group-item">{{ story.title }}</li>
        }
      </ul>
    </div>
    <div class="col">
      <h2>Non-Favourited Stories</h2>
      <ul class="list-group">
        @for (story of nonFavouritedStories$ | async; track story.title) {
        <li class="list-group-item">{{ story.title }}</li>
        }
      </ul>
    </div>
  </div>
</div>
