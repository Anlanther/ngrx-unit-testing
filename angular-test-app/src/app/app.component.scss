.header {
  padding-top: 100px;
}

.content {
  padding-top: 50px;
}
