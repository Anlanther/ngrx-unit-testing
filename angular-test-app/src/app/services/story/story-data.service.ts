import { Injectable } from '@angular/core';
import { Observable, map } from 'rxjs';
import { Story } from '../../models/story.model';
import { DataService } from '../data.service';
import { StoryCollection } from '../models/user-collection.model';

@Injectable({
  providedIn: 'root',
})
export class StoryDataService {
  constructor(private dataService: DataService) {}

  getStories(): Observable<Story[]> {
    const queryName = `getStories`;

    return this.dataService.graphql<Story>(this.getQuery(queryName)).pipe(
      map((response) => {
        return response.appStories.data.map(
          (story: { attributes: StoryCollection }) => ({ ...story.attributes })
        );
      })
    );
  }

  private getQuery(queryName: string, settings?: string) {
    return `query ${queryName}{
      appStories ${settings ?? ''}{
        data {
          attributes {
            title
            author
          }
        }
      }
    }
    `;
  }
}
