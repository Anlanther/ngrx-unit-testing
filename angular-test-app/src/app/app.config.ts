import { ApplicationConfig } from '@angular/core';
import { provideRouter } from '@angular/router';

import { provideEffects } from '@ngrx/effects';
import { provideStore } from '@ngrx/store';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { routes } from './app.routes';
import { AppFeature } from './state/app.state';

import { provideHttpClient } from '@angular/common/http';
import { AppEffects } from './state/app.effect';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideHttpClient(),
    provideStore({ [AppFeature.name]: AppFeature.reducer }),
    provideEffects(AppEffects),
    provideStoreDevtools({ name: 'NGRX Unit Test App', maxAge: 15 }),
  ],
};
