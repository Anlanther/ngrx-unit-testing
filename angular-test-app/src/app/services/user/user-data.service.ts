import { Injectable } from '@angular/core';
import { Observable, map } from 'rxjs';
import { Story } from '../../models/story.model';
import { User } from '../../models/user.model';
import { DataService } from '../data.service';
import { StrapiCollection } from '../models/strapi-collection.model';
import { UserCollection } from '../models/user-collection.model';

@Injectable({
  providedIn: 'root',
})
export class UserDataService {
  constructor(private dataService: DataService) {}

  getUsers(): Observable<User[]> {
    const queryName = `getUsers`;

    return this.dataService.graphql<User>(this.getQuery(queryName)).pipe(
      map((response) => {
        return response.appUsers.data.map(
          (user: { attributes: UserCollection }) => {
            const favouriteStories: Story[] =
              user.attributes.favouriteStories.data.map(
                (story: StrapiCollection<Story>) => ({
                  title: story.attributes.title,
                  author: story.attributes.author,
                })
              );
            return { name: user.attributes.name, favouriteStories };
          }
        );
      })
    );
  }

  private getQuery(queryName: string, settings?: string) {
    return `query ${queryName}{
      appUsers ${settings ?? ''}{
        data {
          attributes {
            name
            favouriteStories {
              data {
                attributes {
                  title
                  author
                }
              }
            }
          }
        }
      }
    }
    `;
  }
}
