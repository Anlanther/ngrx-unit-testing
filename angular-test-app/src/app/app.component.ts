import { CommonModule } from '@angular/common';
import { Component, inject } from '@angular/core';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { RouterOutlet } from '@angular/router';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { AppStore } from './models/app-store.model';
import { Story } from './models/story.model';
import { User } from './models/user.model';
import { AppActions } from './state/app.actions';
import { AppFeature } from './state/app.state';

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule, RouterOutlet, ReactiveFormsModule],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss',
})
export class AppComponent {
  store = inject(Store<AppStore>);
  userForm = new FormControl(null);
  userName = '';
  totalStories = 0;

  allUsers$: Observable<User[]>;
  activeUser$: Observable<User | null>;
  favouritedStories$: Observable<Story[]>;
  nonFavouritedStories$: Observable<Story[]>;

  constructor() {
    this.store.dispatch(AppActions.loadApp());

    this.allUsers$ = this.store.select(AppFeature.selectAllUsers);
    this.activeUser$ = this.store.select(AppFeature.selectActiveUser);
    this.favouritedStories$ = this.store.select(
      AppFeature.selectFavouritedStories
    );
    this.nonFavouritedStories$ = this.store.select(
      AppFeature.selectNonFavouritedStories
    );

    this.activeUser$.subscribe((user) => user && (this.userName = user.name));
  }

  userSelected() {
    const validValue = this.userForm.value;

    if (validValue) {
      this.store.dispatch(AppActions.setActiveUser(validValue));
    }
  }
}
