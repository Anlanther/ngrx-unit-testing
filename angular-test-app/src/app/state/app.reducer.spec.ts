import { Story } from '../models/story.model';
import { User } from '../models/user.model';
import { AppActions } from './app.actions';
import { appReducer, AppState, initialState } from './app.reducer';

describe('AppReducer', () => {
  const mockUsers: User[] = [
    {
      name: 'John Doe',
      favouriteStories: [{ title: 'Matilda', author: 'Roald Dahl' }],
    },
    {
      name: 'Jane Doe',
      favouriteStories: [{ title: 'Fantastic Mr. Fox', author: 'Roald Dahl' }],
    },
  ];
  const mockStories: Story[] = [
    {
      title: 'Story1',
      author: 'A1',
    },
    {
      title: 'Story2',
      author: 'A2',
    },
    { title: 'Matilda', author: 'Roald Dahl' },
    { title: 'Fantastic Mr. Fox', author: 'Roald Dahl' },
  ];

  describe(AppActions.getUsersSuccess.type, () => {
    it('should update the state', () => {
      const state: AppState = initialState;
      const action = AppActions.getUsersSuccess(mockUsers);

      const actual = appReducer(state, action);

      expect(actual).toMatchSnapshot();
    });
  });
  describe(AppActions.getStoriesSuccess.type, () => {
    it('should update the state', () => {
      const state: AppState = initialState;
      const action = AppActions.getStoriesSuccess(mockStories);

      const actual = appReducer(state, action);

      expect(actual).toMatchSnapshot();
    });
  });
  describe(AppActions.setActiveUser.type, () => {
    it('should update the state', () => {
      const state: AppState = initialState;
      const mockActiveUser: User = mockUsers[0];
      const action = AppActions.setActiveUser(mockActiveUser.name);

      const actual = appReducer(state, action);

      expect(actual).toMatchSnapshot();
    });
  });
});
