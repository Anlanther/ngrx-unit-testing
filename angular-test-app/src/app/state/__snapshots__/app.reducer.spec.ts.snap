// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppReducer [App] Get Stories Success should update the state 1`] = `
{
  "activeUser": null,
  "allStories": [
    {
      "author": "A1",
      "title": "Story1",
    },
    {
      "author": "A2",
      "title": "Story2",
    },
    {
      "author": "Roald Dahl",
      "title": "Matilda",
    },
    {
      "author": "Roald Dahl",
      "title": "Fantastic Mr. Fox",
    },
  ],
  "allUsers": [],
  "loading": true,
}
`;

exports[`AppReducer [App] Get Users Success should update the state 1`] = `
{
  "activeUser": null,
  "allStories": [],
  "allUsers": [
    {
      "favouriteStories": [
        {
          "author": "Roald Dahl",
          "title": "Matilda",
        },
      ],
      "name": "John Doe",
    },
    {
      "favouriteStories": [
        {
          "author": "Roald Dahl",
          "title": "Fantastic Mr. Fox",
        },
      ],
      "name": "Jane Doe",
    },
  ],
  "loading": true,
}
`;

exports[`AppReducer [App] Set Active User should update the state 1`] = `
{
  "activeUser": null,
  "allStories": [],
  "allUsers": [],
  "loading": true,
}
`;
